### DISCUSSION TITLE
"Require Signed Commits" is too lax

### DISCUSSION THREAD

**Discussion by ritchiey** [community/community#41154] [Code Security]
### Select Topic Area

Product Feedback

### Body

In my organisation, we keep our private GPG keys only on our Yubikeys. That means that to sign a commit, we have to verify user presence by pressing the button.

However, because of the way GitHub works, there seems to be little point to us taking this precaution because GitHub will happily sign any web based edit on our behalf using its own key and then regard that commit as verified. So the branch protection rule "Require signed commits" only means that the commit was performed by someone with access to login to GitHub as one of our developers.

Anyone that can hijack a GitHub session or remotely compromise a developer computer can perform commits to our repository. Therefore, there seems to be no point in requiring the Yubikey and we may as well keep our private keys on our computer because any attacker who could access the private key on the computer could equally just create a signed commit via GitHub.

One solution to this would be to have an organisational level setting that prevents GitHub from signing commits for edits made via the web interface (or API). Signing merge-commits is (I think) ok, so long as the commits being merged are signed.

**Comment by DNin01** [comment]
I don't think this would prevent a hacker from generating a new GPG key and adding it to a victim's account so they can use it to get through the branch protection, which is just as bad. The only thing GitHub could do about that is make it possible to choose more secure "primary" authentication methods, such as security keys, but we're not there yet.

**Reply by DNin01** [reply]
If you want to protect yourself against session hijacking, a good strategy is to sign out whenever you're done using GitHub for the day. Signing back in will become more seamless than ever when GitHub implements [passkeys](https://fidoalliance.org/passkeys/).
